From 4872dbcdbc8b037ea9764d926825e1f2db41894e Mon Sep 17 00:00:00 2001
From: Simone Rota <simone.rota16@gmail.com>
Date: Thu, 29 Oct 2026 23:47:09 +0000
Subject: [PATCH 08/19] fix(FuncRetToRef): return pointer of variadic functions

The return pointer is the last fixed parameter of the _ret function, but at
the call sites it was appended after the variadic arguments: the module was
broken ("Call parameter type does not match function signature").
---
 passes/FuncRetToRef.cpp | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/passes/FuncRetToRef.cpp b/passes/FuncRetToRef.cpp
index 3b46c9e..d7ffad7 100644
--- a/passes/FuncRetToRef.cpp
+++ b/passes/FuncRetToRef.cpp
@@ -153,6 +153,10 @@ void FuncRetToRef::updateFunctionCalls(Function &Fn, Function &NewFn) {
             args.push_back(V);
         }
 
+        // The pointer to the return value is the last fixed parameter: in a
+        // call to a variadic function it precedes the variadic arguments
+        auto RetArgPos = args.begin() + Fn.getFunctionType()->getNumParams();
+
         IRBuilder<> B(CInstr);
         
         // The function return value can be used immediately by a store instruction, 
@@ -167,7 +171,7 @@ void FuncRetToRef::updateFunctionCalls(Function &Fn, Function &NewFn) {
                 Value *CandidateAlloca = SI->getPointerOperand();
                 if (isa<AllocaInst>(CandidateAlloca)) {
                     found = 1;
-                    args.push_back(SI->getPointerOperand());
+                    args.insert(RetArgPos, SI->getPointerOperand());
                         // remove the store from the basic block as the operation is performed in the called fun
                     SI->eraseFromParent();
                     
@@ -195,7 +199,7 @@ void FuncRetToRef::updateFunctionCalls(Function &Fn, Function &NewFn) {
         if (!found) {
             IRBuilder<> BInit(&(CInstr->getParent()->getParent()->front().front()));
             Instruction *TmpAlloca = BInit.CreateAlloca(CInstr->getType());
-            args.push_back(TmpAlloca);
+            args.insert(args.begin() + Fn.getFunctionType()->getNumParams(), TmpAlloca);
             // do the call
             if (isa<CallInst>(CInstr)) {
                 B.CreateCall(NewFn.getFunctionType(), &NewFn, args);
-- 
2.43.0

